Wikidata-genre UI implementation

pbrackin opened the pull request:
Adding UI for wikidata-genre plugin:
`wikidata.py -> __init__.py, `
adding QT UI files.  
bump version to 1.3.

Plus changes to handle the options (including static and dynamic changes) under various scenarios. 

One setting default I wasn't sure what to do on was the default delimiter.  I left the default delimiter to "; ", but was wondering if you wanted to change the default to the Kodi default of " / "?  Seems there is no standard, sadly, but I brought it up because I noted the relationship between Kodi and Musicbrainz.  But...

pbrackin committed:
Adding UI for wikidata-genre plugin.  wikidata.py -> __init__.py, adding QT UI files.  Bump version to 1.3.

pbrackin committed:
Update __init__.py

add handler for delimiter parsing when expanding the genre list & use the updated list (not the old one!)

phw commented:
Thanks a lot for this, these options are really useful. I currently can't give this a full review (and I hope somebody else comes along for this ;) ), but can you address the small code style issues found by Codacy (https://app.codacy.com/app/MetaBrainz/picard-plugins/pullRequest?prid=2756241)?

From just quickly skimming over it and looking at the available options this looks good to me.

pbrackin committed:
add ability to exclude certain artists from genre pull, clean-up & optimize

New feature: ability to selectively exclude artist-genre-sourcing according to artist.  Also removed reference to global cache, which was a C&P artifact from lastfm plugin, which I absconded some code from.   Also I optimized the RegEx parsers a bit to keep max speed.

pbrackin commented:
I put the PR in too soon... We got some help testing plus I added another config so letting it cook some here is prolly best.  I think no rush on this one anyway.  One of the codacy bugs I was able to clean out, but the other was from the QT gen'd code.  I will have to look at see why its adding an un-required import.  I could change the gen'd code but usually like to avoid that approach.

phw commented:
We can ignore the At generated issue. In Picard itself we have some codacy config to ignore the generated files, we probably should add that rule to the plugins repo as well.

pbrackin committed:
Disable genre delimiter setting if Tags setting is ID3v24, handle & remove blank delimiter, clean up layout, add cutoffs based on settings

Disable genre delimiter setting if Tags setting is ID3v24
Handle blank delimiters
Remove blank delimiter as an option
Clean up options page layout
Add cutoffs to process_track based on settings, to prevent uneeded MB API calls

pbrackin committed:
change "type" parm to "item_type", check opposite case on id3v23 state during options init, some better QT var names, minor house cleaning

change "type" parm to "item_type",
check opposite case on id3v23 state during options init,
some better QT var names,
minor house cleaning for PEP compliance

pbrackin commented:
I think this PR is ready now.  We had several users testing it out over the last few weeks & prior to the last commit I did which was mainly just cosmetic.   If there is anything I can do to help get it reviewed, let me know.  Also, with the success of this "wikidata options UI" feature, some suggestions came in for some additional features which might further the usefulness of this plugin, but I was thinking to address those in a separate ticket.  My work here is probably done for now as my free time resources are getting sucked up by other demands ATM...

phw commented on plugins/wikidata/__init__.py:
```diff
@@ -0,0 +1,397 @@
+            self.ui.genre_delimiter_label.setEnabled(True);
+
+    def info(self):
+        pass
```
Overriding info() should actually not be necessary, since no other plugin providing an options page does this.

The method actually exists in `OptionsPage`, but it's unused and should probably be removed. I currently can't remember for what we originally had this, but it's likely unused for a long time. I'll submit a PR to remove this from Picard, see https://tickets.metabrainz.org/browse/PICARD-1459

phw replied:
This method has now been removed from Picard in https://github.com/metabrainz/picard/pull/1090

pbrackin replied:
OK good; I think this was just brought in as a pattern that existed in another addon.  I will remove this.

phw commented on plugins/wikidata/__init__.py:
```diff
@@ -0,0 +1,397 @@
+
+            for metadata in self.requests[item_id]:
+                if genre_source_type == Wikidata.RELEASE_GROUP:
+                    metadata['release_group_genre_sourced'] = True
```
Is this intentionally a visible variable? This will cause it to be written to tags.

I think this would be better set as a hidden variable `metadata['~release_group_genre_sourced']`. That way it does not show up in file tags, but the user can still use it in scripting. They even can set it as a proper tag again if they wish to see it in their file tags.

pbrackin replied:
Its been a little while, but I can't think of any reason why it would need to be visible.   I didn't know about the private-prefix tilde thingy tho.  I will make it private, I think it should be fine like that.

phw reviewed (changes requested):
I think this is mostly ready for merging, just some minor points up for discussion.

phw commented on plugins/wikidata/__init__.py:
```diff
@@ -0,0 +1,397 @@
+# terms of the Do What The Fuck You Want To Public License, Version 2,
+# as published by Sam Hocevar. See http://www.wtfpl.net/ for more details.
+
+PLUGIN_NAME = 'wikidata-genre'
```
Side question: What do you think about changing the name to "Wikidata Genre"? The current naming really looks odd compated to all other plugins listed.

pbrackin replied:
Yeah its goofy.  When I first saw it, I wondered why they went with a lower-case "w".  But I eventually figured out that the actual wikidata project used a lower case "w".   The "-" is still a mystery to me tho. 

My vote is def FOR congruency in the plugin names, but here's some thoughts... Would changing PLUGIN_NAME break the upgrade path?  As in, peeps with existing wikidata-genre installs wouldn't see the update (somehow they'd have to know to delete the old and start a-fresh with the new name)?  Or worse probably, could they end up with 2 versions of the plugin trying to run at the same time?  Maybe your server has a way to suss that out, or create a special case?

phw replied:
No, wouldn't break anything. This luckily is just the display name. The internal name which is used for updates is "wikidata" (the folder name or in case of a single file plugin the filename without extension)

phw replied:
I went ahead and pushed a commit to change the name to "Wikidata Genre"

pbrackin committed:
remove def info(), make release_group_genre_sourced private

phw committed:
wikidata: Renamed to Wikidata Genre

phw commented:
Merged, thanks a lot for these changes.
